File: src/ua/edu/odeku/ceem/mapRadar/layers/geoName/GeoNamesChunk.java
/*
 * Odessa State environmental University
 * Copyright (C) 2013
 */

package ua.edu.odeku.ceem.mapRadar.layers.geoName;

import gov.nasa.worldwind.geom.Position;
import gov.nasa.worldwind.render.DrawContext;
import gov.nasa.worldwind.render.GeographicText;
import gov.nasa.worldwind.render.UserFacingText;
import ua.edu.odeku.ceem.mapRadar.db.model.GeoName;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * User: Aleo skype: aleo72
 * Date: 19.11.13
 * Time: 14:58
 */
public class GeoNamesChunk {

    protected final GeoNames geoNames;
    protected List<GeoName> geoNameList;

    public GeoNamesChunk(GeoNames geoNames) {
        this.geoNames = geoNames;
        geoNameList = getGeoNameList();
    }

    public List<GeoName> getGeoNameList(){
        if (geoNameList == null){
            geoNameList = Arrays.asList(ua.edu.odeku.ceem.mapRadar.db.model.GeoNames.list(null, geoNames.geoNameCountry, geoNames.geoNameClass, geoNames.geoNameCode));
        }
        return geoNameList;
    }

    public Iterable<GeographicText> makeIterable(DrawContext dc) {
        double maxDisplayDistance = this.geoNames.getMaxDisplayDistance();
        ArrayList<GeographicText> list = new ArrayList<GeographicText>();

        for(GeoName geoName : getGeoNameList()){
            CharSequence str = getText(geoName);
            Position pos = getPosition(geoName);
            GeographicText text = new UserFacingText(str, pos);
            text.setFont(this.geoNames.font);
            text.setColor(this.geoNames.getColor());
            text.setBackgroundColor(this.geoNames.getBackgroundColor());
            text.setVisible(GeoNameLayer.isNameVisible(dc, this.geoNames, pos));
            text.setPriority(maxDisplayDistance);
            list.add(text);
        }
        return list;
    }

    private Position getPosition(GeoName geo) {
        return Position.fromDegrees(geo.lat(), geo.lon(), 0);
    }

    private CharSequence getText(GeoName geoName) {
        if(geoName.translateName() != null && !geoName.translateName().isEmpty()){
            return geoName.name() + "(" + geoName.translateName() + ")";
        } else {
            return geoName.name();
        }
    }
}

File: src/ua/edu/odeku/ceem/mapRadar/layers/geoName/GeoNamesChunk.scala
/*
 * Odessa State environmental University
 * Copyright (C) 2014
 */

package ua.edu.odeku.ceem.mapRadar.layers.geoName

import gov.nasa.worldwind.geom.Position
import gov.nasa.worldwind.render.{GeographicText, UserFacingText, DrawContext}
import ua.edu.odeku.ceem.mapRadar.db.model.GeoName
import ua.edu.odeku.ceem.mapRadar.utils.CeemUtilsFunctions.stringToBoolean

import scala.collection.mutable.ArrayBuffer

/**
 * User: Aleo Bakalov
 * Date: 10.07.2014
 * Time: 13:35
 */
class GeoNamesChunk(val geoNames: GeoNames) {

	val geoNameList = ua.edu.odeku.ceem.mapRadar.db.model.GeoNames.list(null, geoNames.country, geoNames.geoClass, geoNames.geoCode)

	def iterable(dc: DrawContext) = {
		val buff = new ArrayBuffer[GeographicText]()
		for(geoName <- geoNameList){
			val nameGeo = GeoNamesChunk.mkText(geoName)
			val pos = GeoNamesChunk.mkPosition(geoName)
			val text = new UserFacingText(nameGeo, pos)
			text.setFont(this.geoNames.font)
			text.setColor(this.geoNames.color)
			text.setBackgroundColor(this.geoNames.backgroundColor)
			text.setVisible(GeoNameLayer.isNameVisible(dc,this.geoNames, pos))
			text.setPriority(this.geoNames.maxDisplayDistance)
			buff += text
		}
		buff.toList
	}
}

object GeoNamesChunk {
	def mkText(geo: GeoName) = if (geo.translateName.isDefined) geo.name + '(' + geo.translateName + ')' else geo.name

	def mkPosition(geo: GeoName) = Position.fromDegrees(geo.lat, geo.lon, 0)
	
	

}
